storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3MultiPartOutputStream.java

```diff
@@ -86,54 +86,73 @@ public void write(final byte[] b, final int off, final int len) throws IOExcepti
-            // TODO: get rid of this array copying
-            partBuffer.put(source.array(), offset, transferred);
-            source.position(source.position() + transferred);
-            if (!partBuffer.hasRemaining()) {
-                flushBuffer(0, partSize);
+        try {
+            final ByteBuffer source = ByteBuffer.wrap(b, off, len);
+            while (source.hasRemaining()) {
+                final int transferred = Math.min(partBuffer.remaining(), source.remaining());
+                final int offset = source.arrayOffset() + source.position();
+                // TODO: get rid of this array copying
+                partBuffer.put(source.array(), offset, transferred);
+                source.position(source.position() + transferred);
+                if (!partBuffer.hasRemaining()) {
+                    flushBuffer(0, partSize);
+                }
             }
+        } catch (final Exception e) {
+            log.error("Failed to write to stream on upload {}, aborting transaction", uploadId, e);
+            abortUpload();
+            throw new IOException(e);
         }
     }
 
     @Override
     public void close() throws IOException {
-        if (partBuffer.position() > 0) {
-            flushBuffer(partBuffer.arrayOffset(), partBuffer.position());
-        }
-        if (Objects.nonNull(uploadId)) {
+        if (!closed) {
+            if (partBuffer.position() > 0) {
+                try {
+                    flushBuffer(partBuffer.arrayOffset(), partBuffer.position());
+                } catch (final Exception e) {
+                    log.error("Failed to upload last part {}, aborting transaction", uploadId, e);
+                    abortUpload();
+                    throw new IOException(e);
+                }
+            }
             if (!partETags.isEmpty()) {
                 try {
-                    final CompleteMultipartUploadRequest request =
-                        new CompleteMultipartUploadRequest(bucketName, key, uploadId, partETags);
-                    final CompleteMultipartUploadResult result = client.completeMultipartUpload(request);
-                    log.debug("Completed multipart upload {} with result {}", uploadId, result);
+                    completeUpload();
+                    log.debug("Completed multipart upload {}", uploadId);
                 } catch (final Exception e) {
                     log.error("Failed to complete multipart upload {}, aborting transaction", uploadId, e);
-                    client.abortMultipartUpload(new AbortMultipartUploadRequest(bucketName, key, uploadId));
+                    abortUpload();
+                    throw new IOException(e);
                 }
             } else {
-                client.abortMultipartUpload(new AbortMultipartUploadRequest(bucketName, key, uploadId));
+                abortUpload();
             }
         }
+    }
+
+    public boolean isClosed() {
+        return closed;
```

Comment:
If I replace this with just true the tests are passing. Should we use this method inside the class also instead of field directly?

Reply:
Good catch. Using this method on conditions